atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/CoalescingRequestSupplier.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.autobatch;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.google.common.base.Preconditions;
+
+public final class CoalescingRequestSupplier<T> implements CoalescingRequestFunction<Autobatchers.SupplierKey, T> {
```

Comment:
OK. So the main difference I can see between this and `CoalescingSupplier` is that in `CoalescingSupplier` one of the caller threads runs the operation, while with this one we delegate to the specific autobatcher thread. There are pros and cons - this may give better visibility into what work is being done (e.g. if one jstacks), though I'd guess that the other way has less overhead if the operation isn't called very often (since the caller thread will just run the operation without any JVM context switches).

I'm not sure we really need this over `CoalescingSupplier` (as you question in the comments), though if we think the visibility is useful / there's something I'm missing, would be good to clarify that.

Reply:
jstacks would be equivalent in either case. Work would be done in the autobatcher thread, and the caller thread would be blocked waiting on the future from autobatcher. Whereas in `CoalescingSupplier` it would wait on the result of the internal future.

My only concern that comes to mind from when we were doing the timelock stability stuff is the `fairLock`, which seems to be essential to get the correctness guarantees of `CoalescingSupplier` but seems too coarse. Initially we had a coalescing supplier that did the work of `blockOnBecomingLeader`. When I moved it to autobatchers, it was markedly better. The only reasoning that made sense was that threads were trying to wake up to respond to the fact that a result had been published, but because they weren't in order, they'd go back to sleep again waiting to woken up in the correct order. This then sucks because we have one result that needs to be shared across n callers, but they're effectively sorting amongst themselves. More than happy to be wrong here, but I couldn't explain the difference any other way unfortunately.

Initially indifferent, however the above concern shifts it in favour of using autobatcher.

Reply:
Ah, I didn't see the *fair* lock - that makes sense then, it usually imposes quite a bit of a performance overhead. The disruptor-based concurrency model seems like it would work better then.